File: typescript/src/rate.ts
import { parseAmountMinor } from "./amount.js";
import { formatRfc3339Millis, parseRfc3339Millis } from "./time.js";

export interface RationalRateResult {
  readonly amountMinor: string;
}

/** Apply a positive rational rate to a non-negative amount using bankers' rounding. */
export function multiplyRationalHalfEven(
  amountMinor: string,
  numerator: string,
  denominator: string,
): RationalRateResult {
  const amount = parseAmountMinor(amountMinor);
  const parsedNumerator = parseAmountMinor(numerator);
  const parsedDenominator = parseAmountMinor(denominator);
  if (amount < 0n || parsedNumerator <= 0n || parsedDenominator <= 0n) {
    throw new Error(
      "Value rate requires a non-negative amount and positive rational terms",
    );
  }
  const product = amount * parsedNumerator;
  const quotient = product / parsedDenominator;
  const remainder = product % parsedDenominator;
  const doubled = remainder * 2n;
  const rounded =
    doubled > parsedDenominator ||
    (doubled === parsedDenominator && quotient % 2n !== 0n)
      ? quotient + 1n
      : quotient;
  return Object.freeze({
    amountMinor: parseAmountMinor(rounded.toString()).toString(),
  });
}

export interface ValueRateSnapshotInput {
  readonly snapshotId: string;
  readonly baseAsset: string;
  readonly quoteAsset: string;
  readonly numerator: string;
  readonly denominator: string;
  readonly observedAt: string;
  readonly recordedAt: string;
  readonly effectiveAt: string;
  readonly maxStalenessSeconds: number;
}

export interface ValueRateSnapshot extends ValueRateSnapshotInput {
  readonly expiresAt: string;
}

/** Validate and canonicalize a deterministic rational rate observation. */
export function createValueRateSnapshot(
  input: ValueRateSnapshotInput,
): ValueRateSnapshot {
  if (
    input.snapshotId.trim().length === 0 ||
    input.baseAsset.trim().length === 0 ||
    input.quoteAsset.trim().length === 0
  ) {
    throw new Error("Value rate snapshot identity and assets are required");
  }
  if (input.baseAsset === input.quoteAsset) {
    throw new Error("Value rate snapshot requires distinct assets");
  }
  if (
    parseAmountMinor(input.numerator) <= 0n ||
    parseAmountMinor(input.denominator) <= 0n
  ) {
    throw new Error("Value rate numerator and denominator must be positive");
  }
  if (
    !Number.isSafeInteger(input.maxStalenessSeconds) ||
    input.maxStalenessSeconds <= 0
  ) {
    throw new Error(
      "Value rate maximum staleness must be a positive safe integer",
    );
  }
  const observedAt = parseRfc3339Millis(
    input.observedAt,
    "Value rate observedAt",
  );
  const recordedAt = parseRfc3339Millis(
    input.recordedAt,
    "Value rate recordedAt",
  );
  const effectiveAt = parseRfc3339Millis(
    input.effectiveAt,
    "Value rate effectiveAt",
  );
  if (observedAt > recordedAt || recordedAt > effectiveAt) {
    throw new Error(
      "Value rate observed, recorded, and effective times are out of order",
    );
  }
  const expiresAt = effectiveAt + input.maxStalenessSeconds * 1_000;
  if (!Number.isSafeInteger(expiresAt)) {
    throw new Error("Value rate expiry time is outside the supported range");
  }
  return Object.freeze({
    ...input,
    numerator: parseAmountMinor(input.numerator).toString(),
    denominator: parseAmountMinor(input.denominator).toString(),
    expiresAt: formatRfc3339Millis(expiresAt, "Value rate expiresAt"),
  });
}

export type ValueRateFreshness =
  | {
      readonly status: "fresh";
      readonly retainedSnapshotId: string;
      readonly refreshRequired: false;
    }
  | {
      readonly status: "stale";
      readonly retainedSnapshotId: string;
      readonly refreshRequired: true;
    };

/** Determine whether a retained rate observation has crossed its refresh interval. */
export function evaluateValueRateFreshness(input: {
  readonly snapshotId: string;
  readonly capturedAt: string;
  readonly evaluatedAt: string;
  readonly refreshIntervalSeconds: number;
}): ValueRateFreshness {
  if (input.snapshotId.trim().length === 0)
    throw new Error("Value rate snapshot identity is required");
  if (
    !Number.isSafeInteger(input.refreshIntervalSeconds) ||
    input.refreshIntervalSeconds <= 0
  ) {
    throw new Error(
      "Value rate refresh interval must be a positive safe integer",
    );
  }
  const capturedAt = parseRfc3339Millis(
    input.capturedAt,
    "Value rate capturedAt",
  );
  const evaluatedAt = parseRfc3339Millis(
    input.evaluatedAt,
    "Value rate evaluatedAt",
  );
  const ageMillis = evaluatedAt - capturedAt;
  if (ageMillis < 0)
    throw new Error("Value rate evaluation cannot precede its snapshot");
  return ageMillis >= input.refreshIntervalSeconds * 1_000
    ? Object.freeze({
        status: "stale",
        retainedSnapshotId: input.snapshotId,
        refreshRequired: true,
      })
    : Object.freeze({
        status: "fresh",
        retainedSnapshotId: input.snapshotId,
        refreshRequired: false,
      });
}

File: rust/src/rate.rs
use crate::amount::parse_amount_minor;
use crate::time::{format_rfc3339_millis, parse_rfc3339_millis};
use crate::{ValueError, ValueResult, required};

const MAX_RFC3339_MILLIS: i64 = 8_640_000_000_000_000;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RationalRateResult {
    pub amount_minor: String,
}

/// Apply a positive rational rate to a non-negative amount using bankers' rounding.
///
/// # Errors
/// Returns [`ValueError`] when an amount is malformed, a rational term is not positive, or the
/// rounded result exceeds the signed 64-bit range.
pub fn multiply_rational_half_even(
    amount_minor: &str,
    numerator: &str,
    denominator: &str,
) -> ValueResult<RationalRateResult> {
    let amount = parse_amount_minor(amount_minor)?;
    let numerator = parse_amount_minor(numerator)?;
    let denominator = parse_amount_minor(denominator)?;
    if amount < 0 || numerator <= 0 || denominator <= 0 {
        return Err(ValueError::new(
            "Value rate requires a non-negative amount and positive rational terms",
        ));
    }
    let product = i128::from(amount) * i128::from(numerator);
    let denominator = i128::from(denominator);
    let quotient = product / denominator;
    let remainder = product % denominator;
    let doubled = remainder * 2;
    let rounded = if doubled > denominator || (doubled == denominator && quotient % 2 != 0) {
        quotient + 1
    } else {
        quotient
    };
    Ok(RationalRateResult {
        amount_minor: i64::try_from(rounded)
            .map_err(|_| ValueError::new("Value amount exceeds the supported signed range"))?
            .to_string(),
    })
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ValueRateSnapshotInput {
    pub snapshot_id: String,
    pub base_asset: String,
    pub quote_asset: String,
    pub numerator: String,
    pub denominator: String,
    pub observed_at: String,
    pub recorded_at: String,
    pub effective_at: String,
    pub max_staleness_seconds: i64,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ValueRateSnapshot {
    pub snapshot_id: String,
    pub base_asset: String,
    pub quote_asset: String,
    pub numerator: String,
    pub denominator: String,
    pub observed_at: String,
    pub recorded_at: String,
    pub effective_at: String,
    pub max_staleness_seconds: i64,
    pub expires_at: String,
}

/// Validate and canonicalize a deterministic rational rate observation.
///
/// # Errors
/// Returns [`ValueError`] when identities, rational terms, timestamps, or staleness are invalid.
pub fn create_value_rate_snapshot(input: ValueRateSnapshotInput) -> ValueResult<ValueRateSnapshot> {
    required(
        &input.snapshot_id,
        "Value rate snapshot identity and assets are required",
    )?;
    required(
        &input.base_asset,
        "Value rate snapshot identity and assets are required",
    )?;
    required(
        &input.quote_asset,
        "Value rate snapshot identity and assets are required",
    )?;
    if input.base_asset == input.quote_asset {
        return Err(ValueError::new(
            "Value rate snapshot requires distinct assets",
        ));
    }
    let numerator = parse_amount_minor(&input.numerator)?;
    let denominator = parse_amount_minor(&input.denominator)?;
    if numerator <= 0 || denominator <= 0 {
        return Err(ValueError::new(
            "Value rate numerator and denominator must be positive",
        ));
    }
    if input.max_staleness_seconds <= 0 {
        return Err(ValueError::new(
            "Value rate maximum staleness must be a positive integer",
        ));
    }
    let observed = parse_rfc3339_millis(&input.observed_at, "Value rate observedAt")?;
    let recorded = parse_rfc3339_millis(&input.recorded_at, "Value rate recordedAt")?;
    let effective = parse_rfc3339_millis(&input.effective_at, "Value rate effectiveAt")?;
    if observed > recorded || recorded > effective {
        return Err(ValueError::new(
            "Value rate observed, recorded, and effective times are out of order",
        ));
    }
    let expires = input
        .max_staleness_seconds
        .checked_mul(1_000)
        .and_then(|staleness| effective.checked_add(staleness))
        .ok_or_else(|| ValueError::new("Value rate expiry time is outside the supported range"))?;
    if !(-MAX_RFC3339_MILLIS..=MAX_RFC3339_MILLIS).contains(&expires) {
        return Err(ValueError::new(
            "Value rate expiry time is outside the supported RFC 3339 calendar range",
        ));
    }
    Ok(ValueRateSnapshot {
        snapshot_id: input.snapshot_id,
        base_asset: input.base_asset,
        quote_asset: input.quote_asset,
        numerator: numerator.to_string(),
        denominator: denominator.to_string(),
        observed_at: input.observed_at,
        recorded_at: input.recorded_at,
        effective_at: input.effective_at,
        max_staleness_seconds: input.max_staleness_seconds,
        expires_at: format_rfc3339_millis(expires),
    })
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ValueRateFreshness {
    Fresh {
        retained_snapshot_id: String,
        refresh_required: bool,
    },
    Stale {
        retained_snapshot_id: String,
        refresh_required: bool,
    },
}

/// Determine whether a retained rate observation has crossed its refresh interval.
///
/// # Errors
/// Returns [`ValueError`] when the identity, interval, or timestamps are invalid.
pub fn evaluate_value_rate_freshness(
    snapshot_id: &str,
    captured_at: &str,
    evaluated_at: &str,
    refresh_interval_seconds: i64,
) -> ValueResult<ValueRateFreshness> {
    required(snapshot_id, "Value rate snapshot identity is required")?;
    if refresh_interval_seconds <= 0 {
        return Err(ValueError::new(
            "Value rate refresh interval must be a positive integer",
        ));
    }
    let captured = parse_rfc3339_millis(captured_at, "Value rate capturedAt")?;
    let evaluated = parse_rfc3339_millis(evaluated_at, "Value rate evaluatedAt")?;
    let age = evaluated
        .checked_sub(captured)
        .ok_or_else(|| ValueError::new("Value rate age is outside the supported range"))?;
    if age < 0 {
        return Err(ValueError::new(
            "Value rate evaluation cannot precede its snapshot",
        ));
    }
    let interval = refresh_interval_seconds.checked_mul(1_000).ok_or_else(|| {
        ValueError::new("Value rate refresh interval is outside the supported range")
    })?;
    Ok(if age >= interval {
        ValueRateFreshness::Stale {
            retained_snapshot_id: snapshot_id.to_owned(),
            refresh_required: true,
        }
    } else {
        ValueRateFreshness::Fresh {
            retained_snapshot_id: snapshot_id.to_owned(),
            refresh_required: false,
        }
    })
}
